### ISSUE TITLE
data.google_client_config issue in 4.65

### ISSUE THREAD

**Issue by srlightbody** [hashicorp/terraform-provider-google#14602]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

1.4.6

### Affected Resource(s)

* google_client_config
* google_container_cluster

### Terraform Configuration Files

```tf

data "google_client_config" "cluster" {
  provider = google
}

data "google_container_cluster" "region-cluster" {
  project  = terraform.workspace
  name     = "${replace(terraform.workspace, "foo-", "")}-regop"
  location = "us-region"
}

provider "kubernetes" {
  alias                  = "us-region"
  host                   = data.google_container_cluster.west1-cluster.endpoint != null ? "https://${data.google_container_cluster.region-cluster.endpoint}" : null
  token                  = data.google_client_config.cluster.access_token
  cluster_ca_certificate = data.google_container_cluster.region-cluster.master_auth != null ? base64decode(data.google_container_cluster.region-cluster.master_auth[0].cluster_ca_certificate, ) : null
}
```

### Debug Output

All kubernetes resources relying on the above auth snippet start failing. They all report errors similar to - 
```
Error: namespaces "default" is forbidden: User "system:anonymous" cannot get resource "namespaces" in API group "" in the namespace "default"
Error: services "foo" is forbidden: User "system:anonymous" cannot get resource "services" in API group "" in the namespace "default"
Error: configmaps "foo" is forbidden: User "system:anonymous" cannot get resource "configmaps" in API group "" in the namespace "default"
```

### Expected Behavior

Resources and authentication to GKE clusters continue to work as normal

### Actual Behavior

The kubernetes provider relying on google data sources google_client_config and google_container_cluster begins failing with errors like the ones listed in the debug output section

### Steps to Reproduce

1. Update google provider to 4.65
2. Run a terraform plan

**Comment by edwardmedia** [comment]
@srlightbody I have just tried your above config, all attributes of these two data sources look fine. I don't see any code changes in the `v4.65.0` that possibly related to this issue.  

One possible reason is that one of apis was experiencing difficulties. Do you still see the issue? If yes, can you detail the steps to repro this issue?

**Comment by yvjessestephens** [comment]
For us, it seems the access_token from data.google_client_config is coming back null. 

`data "google_client_config" "default" {}

data "google_container_cluster" "cluster" {
  name = var.GKE_CLUSTER
  location = var.REGION
}
provider "kubernetes" {
  host  = "https://${data.google_container_cluster.cluster.endpoint}"
  token = "foo${data.google_client_config.default.access_token}"
  cluster_ca_certificate = base64decode(
    data.google_container_cluster.cluster.master_auth[0].cluster_ca_certificate,
  )
}`

I put the token in quotes with some other characters, foo, so I could see what the value was, but you can see that the expression fails with the following error:
╷
│ Error: Invalid template interpolation value
│ 
│   on provider.tf line 39, in provider "kubernetes":
│   39:   token = "foo${data.google_client_config.default.access_token}"
│ 
│ The expression result is null. Cannot include a null value in a string
│ template.
╵

**Comment by brokenjacobs** [comment]
We are seeing the same issue

**Comment by edwardmedia** [comment]
@srlightbody @brokenjacobs possible dup of https://github.com/hashicorp/terraform-provider-google/issues/13950

Does it occur only after v4.65.0?

**Comment by shuyama1** [comment]
Thanks for confirming the issue. I suspect the issue may be related to the change in https://github.com/GoogleCloudPlatform/magic-modules/pull/7875. I opened a revert PR https://github.com/GoogleCloudPlatform/magic-modules/pull/7948 while investigating the root cause and finding a fix forward. The presubmit tests take ~1.5 hours to run.

**Comment by srlightbody** [comment]
It does occur only on 4.65 for us. I'm still able to repro the issue, all I'm doing to cause it is switch a repo using the snippet I listed above over to using 4.65 and then running a plan. The plan will fail with those errors consistently. As soon as I limit the max version to 4.64, things go back to normal.

**Comment by brokenjacobs** [comment]
I just pinned us back to 4.64.0 and the problem went away, so I'd say this is definitely a 4.65.0 issue

**Comment by gabClark** [comment]
Similarly experiencing issue with `google_client_config` we depend on it in our setup to determine project and since 4.65 it fails with : `data.google_client_config.current.project is null` . Pinning to 4.64 temporarily.

**Comment by philip-harvey** [comment]
for us blocking 4.65 works around the issue.
e.g.
version = "~> 4.40, !=4.65"  # 4.65 is broken. https://github.com/hashicorp/terraform-provider-google/issues/14602

**Comment by shuyama1** [comment]
Please pin to the previous version before we have a fix out. Could someone provide with which related environment variables/ credentials are set for provider configuration?

**Comment by cmaahs** [comment]
Same as [yvjessestephens](https://github.com/yvjessestephens) reported.  

Provider spec looks like so

```
provider "google" {
  project = local.state_project
  region  = local.state_region
}
```

The `data` references a different gcp project than the base provider configuration.

**Comment by m-denton** [comment]
Same issue, presented itself on tf applying for k8s services/secrets. Example of errors are seen below. Pinning to the previous version fixed the issue. Here is how my provider is set up (the credentials are set to the json key of a GCP service account).

```
provider "google" {
  credentials = data.vault_generic_secret.google_json.data_json
  project = "b2z-insurance"
  region = "us-central1"
}
```

Errors from apply:
```
Error: query: failed to query with labels: secrets is forbidden: User "system:anonymous" cannot list resource "secrets" in API group "" in the namespace "<ns>"

Error: secrets "<service-name>" is forbidden: User "system:anonymous" cannot get resource "secrets" in API group "" in the namespace "<ns>"

Error: services "<service-name>"  is forbidden: User "system:anonymous" cannot get resource "services" in API group "" in the namespace "<ns>"
```

**Comment by m-denton** [comment]
Same issue, presented itself on tf applying for k8s services/secrets. Example of errors are seen below. Pinning to the previous version fixed the issue. Here is how my provider is set up (the credentials are set to the json key of a GCP service account).

```
provider "google" {
  credentials = data.vault_generic_secret.google_json.data_json
  project = "<project-id>"
  region = "us-central1"
}
```

Errors from apply:
```
Error: query: failed to query with labels: secrets is forbidden: User "system:anonymous" cannot list resource "secrets" in API group "" in the namespace "<ns>"

Error: secrets "<service-name>" is forbidden: User "system:anonymous" cannot get resource "secrets" in API group "" in the namespace "<ns>"

Error: services "<service-name>"  is forbidden: User "system:anonymous" cannot get resource "services" in API group "" in the namespace "<ns>"
```

**Comment by shuyama1** [comment]
`4.65.1` should be out now and may potentially fix the issue. Unfortunately, I am still unable to get a reproduction of the error.  Can someone try upgrading and confirm if the issues is resolved in `4.65.1`?

**Comment by safredAAs** [comment]
Tried with 4.65.1, I still get the error 

Error: Invalid template interpolation value
Authorization = "Bearer ${data.google_client_config.default.access_token}"
|----------------
| data.google_client_config.default.access_token is null
The expression result is null. Cannot include a null value in a string

**Comment by shuyama1** [comment]
Thanks @safredAAs! I finally was able to reproduce the error after unsetting environment variable GOOGLE_ZONE and GOOGLE_REGION. Working on a fix now. Apologize for the issue.

**Comment by gjumic-optiva** [comment]
Seeing a similar issue on v4.65.1 and v4.65.0. Forcing Terraform to use v4.64.0 as a workaround resolves the issue.
```
Error: secrets is forbidden: User "system:anonymous" cannot create resource "secrets" in API group "" in the namespace "kube-system": No policy matched.
Error: secrets is forbidden: User "system:anonymous" cannot create resource "secrets" in API group "" in the namespace "default": No policy matched.
Error: unable to build kubernetes objects from release manifest: unknown
```

**Comment by audriusb** [comment]
it does seem that providing zone env variable helps as a workaround.

**Comment by mahendra03** [comment]
We are also getting below error 

- Installing hashicorp/google v4.65.1...
- Installed hashicorp/google v4.65.1 (signed by HashiCorp)
- Installing hashicorp/google-beta v4.65.1...
- Installed hashicorp/google-beta v4.65.1 (signed by HashiCorp)****

Error: Missing required argument

  with module.gke_deployment.provider["gitrepo/company/kubeconfig"],
  on ../../modules/gke_deployment/providers.tf line 19, in provider "google":
  19:   token = data.google_client_config.creator.access_token

The argument "token" is required, but no definition was found.

When we expect for the fix ?

**Comment by mahendra03** [comment]
> ow and may potentially fix the issue. Unfortunately, I am still unable to get a reproduction of the error. Can someone try upgrading and confirm if the

We are facing same issue on 

nstalling hashicorp/google v4.65.1...
Installed hashicorp/google v4.65.1 (signed by HashiCorp)
Installing hashicorp/google-beta v4.65.1...
Installed hashicorp/google-beta v4.65.1 (signed by HashiCorp)****

**Comment by hakuno** [comment]
+1 same here

I had to pin `4.64.0` back.

**Comment by shuyama1** [comment]
v4.65.2 is release and it should resolve the problem. Please let me know if the error still occurs.

**Comment by shuyama1** [comment]
Closing this ticket now as the issue should be fixed in v4.65.2. Please comment or reopen this issue if the error shows up again.
